### ISSUE TITLE
[feature] Allow signing raw transaction in trezorctl, eg BIP-174 PSBT

### ISSUE THREAD

**Issue by jonathancross** [trezor/python-trezor#183]
The current `sign_tx` command collects info manually from the user, then _generates_ a transaction (which is then signed).
This is an extremely manual and potentially error-prone process.

Would be great if users could sign a raw transaction that was pre-generated in Bitcoin Core (via `createrawtransaction`) or Electrum (via [`payto`](https://github.com/spesmilo/electrum-docs/blob/master/coldstorage_cmdline.rst#create-an-unsigned-transaction) command).

## Benefits
Such a feature would be especially helpful for off-line tx signing (to prevent password from being leaked via malware). The user's online (watch-only) wallet could determine optimal fees, find correct inputs, check destination addresses and amounts, etc. 

Would also be good for multisig transactions that are passed around.  More motivating use case here: https://en.bitcoin.it/wiki/Raw_Transactions#Motivating_use_cases

**Comment by saleemrashid** [comment]
You would still need to manage the metadata for the transaction (BIP32 paths, for example) so it may be just as easy to generate the Protocol Buffers for the transaction (either as Python or serialized as binary) that you could transfer to the airgap and send to the TREZOR.

On the other hand, we could add code for signing a raw transaction with provided metadata. This would especially be useful to replace the "hack" I'm using for Bitcoin Core hardware wallet support.

**Comment by jb55** [comment]
@saleemrashid 

> You would still need to manage the metadata for the transaction (BIP32 paths, for example)

I effectively had to do this manually the other day by generating a list of my changes addresses to find the right index to use in the sign_tx prompt.

I wonder if `sign_raw_tx` could be smart enough to search through your key chains for the correct derivation path. Then you could use something like bitcoin-tx to generate the transaction without signatures, and then pass it to `sign_raw_tx`?

**Comment by saleemrashid** [comment]
@jb55 That would probably take quite a long time. In terms of your suggestion to use `bitcoin-tx`, I have been working on basic hardware wallet support for Bitcoin Core. Bitcoin Core keeps a database of all the addresses to paths in the wallet.

**Comment by jonathancross** [comment]
Providing the BIP32  path (eg `m/44'/0'/0'`) would be fine as long as it could then iterate through to find the necessary signing keys. Are you saying that would take too long?

We'd want to do this anyway so support multiple coin types and accounts.

**Comment by prusnak** [comment]
I agree with Saleem that providing rawtx is not enough and that we have to provide metadata in addition to rawtx.

For format of these metadata we might want to have an input from @jonasschnelli who was trying to design such format for cross-vendor hardware wallet signing.

**Comment by NicolasDorier** [comment]
So for a hardware wallet to be able to sign something here is what is needed.

1. The raw transaction
2. The BIP32 paths to sign
3. If segwit, the amount of the outputs,
4. If not segwit, all the previous transactions.

I think Bitcoin Core would need probably a separate rpc method than `createrawtransaction`.

I thought there was a meeting HW manufacturers like slush/Ledger/Digital Bitbox (@jonasschnelli) back in may 2017 to decide about a standard. (am I wrong?)

Regardless of the existence of a standard of not, I think having a RPC method in bitcoin core giving the above information necessary information to sign offline would be very useful.

**Comment by instagibbs** [comment]
You also need the redeemscript when applicable.

**Comment by saleemrashid** [comment]
@instagibbs Good point. Originally, I was trying to use the output of `decoderawtransaction` and the BIP-32 paths but that was not useful for multisignature. Having the redeem scripts (perhaps the output of `decodescript`) would resolve that.

**Comment by instagibbs** [comment]
If we're discussing adopting a standard for data to pass, I think a good
starting place is BIP174:
https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki

It's probably the most reviewed and complete.

**Comment by instagibbs** [comment]
I have also implemented my own Bitcoin Core integration with Ledger, based off of @saleemrashid 's work.

Implementing BIP174 is on my list of TODO, since it's required for a useful RPC interface for general signing as well.

cc @achow101

**Comment by jb55** [comment]
Gregory Sanders <notifications@github.com> writes:
> I have also implemented my own Bitcoin Core integration with Ledger, based off of @saleemrashid 's work.
>
> Implementing BIP174 is on my list of TODO, since it's required for a useful RPC interface for general signing as well.

I would be interested in helping out with this. Is there a place to
coordinate Bitcoin Core <> HW wallet dev?

**Comment by instagibbs** [comment]
@jb55 it's being done piecemeal by a few people because it's not upstreamable.

The current Core wallet architecture has a notion of `ISMINE` which means if you have the key, it's your money, and if you don't have the key, it's not your money(aside from watchonly, which is another bolt-on system).

sipa has been having thoughts on how to make this better: https://gist.github.com/sipa/125cfa1615946d0c3f3eec2ad7f250a2

Until then. for example, my work explicitly disallows address importing of any sort, because the hack from @saleemrashid 's branch is to say any address in its wallet is something it can sign for.

If you're curious to play around with a non-segwit variant check it out here: https://github.com/instagibbs/bitcoin/tree/external_sign_3

**Comment by NicolasDorier** [comment]
@instagibbs actually, what is needed is the *scriptCode* to be exact. Yeah I missed that one.

**Comment by jonasschnelli** [comment]
Signing raw-transactions is more complicated to it looks on the first sight.
It requires more data then just the raw transaction.

What we need is a standard format for unsigned Bitcoin transactions that have all sufficient data required to verify and sign the transaction. Also, it should respect multi signature use-cases.

IMO the best current proposal is BIP174 https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki
I guess it misses the non-segwit (legacy) inputs (full raw transaction of all legacy inputs).

That standard could even be a file-format (mime-type) so people could send it via Email, etc.

**Comment by prusnak** [comment]
Agree with Jonas. I'll be happy to add such functionality once a standard for partially signed transactions emerges.

**Comment by instagibbs** [comment]
@jonasschnelli it has non-segwit input fields

**Comment by achow101** [comment]
@jonasschnelli @prusnak [BIP 174](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki) should contain everything that you need in order to sign from a hardware wallet or offline machine. I designed it with the main usage being for hardware wallets. It has fields to specify both witness UTXOs (just the output) and non-witness UTXOs (the entire previous transaction), fields to hold signatures before the scriptSig is constructed, and fields to specify the BIP 32 derivation paths of keys to be used. If it is missing anything, please [comment on it](https://github.com/bitcoin/bips/wiki/Comments:BIP-0174) and new fields can be easily added which includes whatever other data that you need.

**Comment by NicolasDorier** [comment]
@achow101 BIP174 does not allow to pass the `scriptCode`. RedeemScript or WitnessRedeemScript are not necessarily what is being signed.

**Comment by achow101** [comment]
@NicolasDorier Can you give an example where the `scriptCode` would not be part of the `redeemScript`, `witnessScript`, or `scriptPubKey`? In such cases, how would the signer know to use that `scriptCode` instead of something else?

**Comment by NicolasDorier** [comment]
@achow101 one example is `OP_CODESEPARATOR` (which I am using for tumblebit on https://github.com/bitcoin/bitcoin/pull/11423) This allow me to have smaller scripts by allowing the same pubkey to sign a specific script execution path.

**Comment by NicolasDorier** [comment]
@achow101 actually for hardware wallet redeemScript and witness script are not needed at all. Only the scriptCode is needed. What to do with the signature should be up to the wallet software, not the hardware wallet.

**Comment by instagibbs** [comment]
@NicolasDorier fwiw your answer is "no", since scriptCode is a subset of the various scripts.

I think I understand your desire for the field though, and will try to explain why it can only be an additional field of some sort, not a replacement or generalization.

The point of the BIP is to give the signer everything it needs to sign/solve any arbitrary script in a standardized way, allowing it to be passed around in any order to completion without loss of information. Your point seems to be that if you try to sign something with for example OP_CODESEPERATOR, with a device that doesn't understand how to sign specific subsets of the redeemScript/witnessScript, none of the fields will match that use-case, assuming you're directly serializing the PSBT and sending it directly to Trezor and it just tries to sign the whole thing.

Only supporting `scriptCode` seems to imply that the host will have to do surgery on existing fields to pass it along to the hww, as the host will likely want the whole script for verification, then remove parts the hww isn't supposed to sign, get the answer back, stitch it back together, then pass it along to the next signer. This is pretty ugly and seems to at least partially defeat the purpose of the standard.

In other words, I think you want a field that signals to a dumb signer(or any signer in general) which subset of the script to sign, only in the case of OP_SEPARATOR.

spitballing solutions on a couple minutes thought:
1) Add a field for `scriptCode`, which is a field intended to be set to signal to signing hosts/hww what part is desired to be signed.
2) Add a field that says how many bytes of the script to skip ahead before signing(implicit scriptCode). Or perhaps how many OP_CODESEPARATORs to process to get to the `scriptCode` desired(more assumed of the signer's ability to understand script).

Either way, I think this is an issue that can be solved later on top of BIP174, unless there is a more fundamental, non-OP_CODESEPARATOR related issue.

**Comment by NicolasDorier** [comment]
> The point of the BIP is to give the signer everything it needs to sign/solve any arbitrary script in a standardized way, allowing it to be passed around in any order to completion without loss of information. 

I understand, and for that, you only need 2 information to cover all use cases: `ScriptCode and HashVersion`. There is no need of any other fields.

It is the approach of Ledger Wallet HW API which works fine for any P2W-(P2SH) combinaison along with any strange OP_CODESEP usecase. (Even if not originally designed for it)

But I think you want the host to be able to verify that the transaction is correctly signed as well.

>  as the host will likely want the whole script for verification

Ah I think this is what I initially did not understood. I was seeing BIP174 only from the point of view of the HW signer, where the signer only goal was : Identifying the change address, Calculating the transaction fees correctly, and signing correctly. (for this, only `ScriptCode + HashVersion` is important)

But the goal of BIP174 goes beyond that.

**Comment by achow101** [comment]
> But the goal of BIP174 goes beyond that.

Yes. The goal is to provide everything that is needed to sign and construct the final transaction.

**Comment by matejcik** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [support raw transactions BIP174](https://trello.com/c/bNVajicD/3-support-raw-transactions-bip174)

**Comment by jonathancross** [comment]
FYI: BIP-174 support [has landed for Bitcoin Core v.0.17.0](https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.17.0.md#bip-174-partially-signed-bitcoin-transactions-support)

**Comment by instagibbs** [comment]
https://github.com/achow101/HWI is a python repo for PSBT.

Would be great if someone could get multisig working for Trezor in the repo.

**Comment by yahyameo** [comment]
What meta data need to pass to sign the transactions?

**Comment by jonathancross** [comment]
@yahyameo See [`trezorctl sign-tx` example](https://github.com/trezor/python-trezor/blob/master/docs/EXAMPLES.rst#transaction-signing).

**Comment by matejcik** [comment]
closing this in favor of https://github.com/trezor/trezor-firmware/issues/38, which is a necessary precondition for true PSBT signing.

simple single-signer PSBT will be supported through microwallet
